File: backend/device_service/api/routes/enrollment.py
"""API routes for Enrollment management."""

from fastapi import APIRouter, HTTPException, status, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from device_service.core.database import get_db
from device_service.services.enrollment_service import EnrollmentService
from device_service.api.dependencies import get_current_user
from shared.schemas.enrollment import (
    EnrollmentStartRequest,
    EnrollmentStartResponse,
    EnrollmentSessionResponse,
    EnrolledFingersResponse,
    EnrollmentRecordSummary,
    EnrollmentListResponse,
)
from shared.schemas.user import UserResponse
from device_service.exceptions import (
    DeviceOfflineError,
    DeviceNotFoundError,
    EnrollmentError,
    EnrollmentInProgressError,
)

router = APIRouter(prefix="/api/v1/enrollment", tags=["enrollment"])


@router.post(
    "/start",
    response_model=EnrollmentStartResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Start enrollment",
    description="""
    Start fingerprint enrollment for a student on a device.
    
    This endpoint:
    1. Validates student and device exist
    2. Checks device is online
    3. Creates enrollment session
    4. Sends enrollment command to device
    5. Returns session information
    
    The device will enter enrollment mode and wait for the student to place their finger.
    Real-time progress updates are sent via WebSocket (to be implemented in Phase 3).
    """,
    responses={
        201: {"description": "Enrollment started successfully"},
        400: {"description": "Invalid request"},
        404: {"description": "Student or device not found"},
        503: {"description": "Device is offline"},
    },
)
async def start_enrollment(
    request: EnrollmentStartRequest,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Start enrollment for a student.
    
    Args:
        request: Enrollment start request with student_id, device_id, and finger_id
        current_user: Authenticated user (automatically injected)
        db: Database session (automatically injected)
        
    Returns:
        EnrollmentStartResponse with session information
    """
    enrollment_service = EnrollmentService(db)
    
    try:
        enrollment_session = await enrollment_service.start_enrollment(
            student_id=request.student_id,
            device_id=request.device_id,
            finger_id=request.finger_id,
            school_id=current_user.school_id,
        )
        
        return EnrollmentStartResponse(
            session_id=enrollment_session.session_id,
            student_id=enrollment_session.student_id,
            device_id=enrollment_session.device_id,
            finger_id=enrollment_session.finger_id,
            status=enrollment_session.status,
            started_at=enrollment_session.started_at,
        )
        
    except DeviceNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except DeviceOfflineError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(e)
        )
    except EnrollmentInProgressError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(e)
        )
    except EnrollmentError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        # Log unexpected errors
        import logging
        logger = logging.getLogger(__name__)
        logger.error(f"Unexpected error starting enrollment: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An unexpected error occurred while starting enrollment"
        )


def _enrollment_to_summary(session) -> EnrollmentRecordSummary:
    """Build EnrollmentRecordSummary from EnrollmentSession."""
    return EnrollmentRecordSummary(
        id=session.id,
        session_id=session.session_id,
        student_id=session.student_id,
        device_id=session.device_id,
        finger_id=session.finger_id,
        quality_score=session.quality_score,
        completed_at=session.completed_at,
        has_template=bool(session.template_data),
        student_name=(
            f"{session.student.first_name} {session.student.last_name}"
            if session.student
            else None
        ),
        device_name=session.device.name if session.device else None,
    )


@router.get(
    "/students/{student_id}/enrollments",
    response_model=EnrollmentListResponse,
    summary="List enrollments by student",
    description="Get completed enrollment records for a student (for UI and sync readiness).",
)
async def list_enrollments_by_student(
    student_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List completed enrollments for a student."""
    enrollment_service = EnrollmentService(db)
    sessions = await enrollment_service.get_enrollments_by_student(
        school_id=current_user.school_id,
        student_id=student_id,
    )
    return EnrollmentListResponse(
        enrollments=[_enrollment_to_summary(s) for s in sessions]
    )


@router.get(
    "/devices/{device_id}/enrollments",
    response_model=EnrollmentListResponse,
    summary="List enrollments by device",
    description="Get completed enrollment records for a device (for UI and sync readiness).",
)
async def list_enrollments_by_device(
    device_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List completed enrollments for a device."""
    enrollment_service = EnrollmentService(db)
    sessions = await enrollment_service.get_enrollments_by_device(
        school_id=current_user.school_id,
        device_id=device_id,
    )
    return EnrollmentListResponse(
        enrollments=[_enrollment_to_summary(s) for s in sessions]
    )


@router.get(
    "/devices/{device_id}/students/{student_id}/fingers",
    response_model=EnrolledFingersResponse,
    summary="Get enrolled fingers",
    description="Get list of finger IDs (0-9) already enrolled for this student on the device.",
)
async def get_enrolled_fingers(
    device_id: int,
    student_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Return enrolled finger IDs for student on device."""
    enrollment_service = EnrollmentService(db)
    try:
        finger_ids = await enrollment_service.get_enrolled_fingers(
            device_id=device_id,
            student_id=student_id,
            school_id=current_user.school_id,
        )
        return EnrolledFingersResponse(
            device_id=device_id,
            student_id=student_id,
            finger_ids=finger_ids,
        )
    except DeviceNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except DeviceOfflineError as e:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(e))


@router.delete(
    "/devices/{device_id}/students/{student_id}/fingers/{finger_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete fingerprint",
    description="Delete the fingerprint template for the given student and finger on the device.",
)
async def delete_fingerprint(
    device_id: int,
    student_id: int,
    finger_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Delete fingerprint from device."""
    if not 0 <= finger_id <= 9:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="finger_id must be 0-9")
    enrollment_service = EnrollmentService(db)
    try:
        await enrollment_service.delete_fingerprint(
            device_id=device_id,
            student_id=student_id,
            finger_id=finger_id,
            school_id=current_user.school_id,
        )
    except DeviceNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except DeviceOfflineError as e:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(e))
    except EnrollmentError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post(
    "/cancel/{session_id}",
    response_model=EnrollmentSessionResponse,
    summary="Cancel enrollment",
    description="""
    Cancel an ongoing enrollment session.
    
    Sends cancel command to the device and broadcasts cancellation via WebSocket
    so the frontend receives the update immediately.
    """,
)
async def cancel_enrollment(
    session_id: str,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Cancel an enrollment session by session_id."""
    enrollment_service = EnrollmentService(db)
    
    try:
        enrollment_session = await enrollment_service.cancel_enrollment_by_session_id(
            session_id=session_id,
            school_id=current_user.school_id,
        )
        
        if not enrollment_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Enrollment session {session_id} not found"
            )
        
        return EnrollmentSessionResponse(
            id=enrollment_session.id,
            session_id=enrollment_session.session_id,
            student_id=enrollment_session.student_id,
            device_id=enrollment_session.device_id,
            finger_id=enrollment_session.finger_id,
            school_id=enrollment_session.school_id,
            status=enrollment_session.status,
            error_message=enrollment_session.error_message,
            template_data=enrollment_session.template_data,
            quality_score=enrollment_session.quality_score,
            started_at=enrollment_session.started_at,
            completed_at=enrollment_session.completed_at,
            created_at=enrollment_session.created_at,
            updated_at=enrollment_session.updated_at,
            is_deleted=enrollment_session.is_deleted,
        )
        
    except EnrollmentError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get(
    "/status/{session_id}",
    response_model=EnrollmentSessionResponse,
    summary="Check enrollment status",
    description="""
    Check the status of an enrollment session.
    
    This endpoint polls the device to check if enrollment has completed.
    Returns the current enrollment session status.
    """,
)
async def check_enrollment_status(
    session_id: str,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Check enrollment status for a session."""
    enrollment_service = EnrollmentService(db)
    
    try:
        enrollment_session = await enrollment_service.repository.get_by_session_id(
            session_id, current_user.school_id
        )
        
        if not enrollment_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Enrollment session {session_id} not found"
            )
        
        return EnrollmentSessionResponse(
            id=enrollment_session.id,
            session_id=enrollment_session.session_id,
            student_id=enrollment_session.student_id,
            device_id=enrollment_session.device_id,
            finger_id=enrollment_session.finger_id,
            school_id=enrollment_session.school_id,
            status=enrollment_session.status,
            error_message=enrollment_session.error_message,
            template_data=enrollment_session.template_data,
            quality_score=enrollment_session.quality_score,
            started_at=enrollment_session.started_at,
            completed_at=enrollment_session.completed_at,
            created_at=enrollment_session.created_at,
            updated_at=enrollment_session.updated_at,
            is_deleted=enrollment_session.is_deleted,
        )
        
    except HTTPException:
        raise
    except Exception as e:
        import logging
        logger = logging.getLogger(__name__)
        logger.error(f"Unexpected error checking enrollment status: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An unexpected error occurred while checking enrollment status"
        )

File: frontend/lib/api/enrollment.ts
/**
 * Enrollment API client.
 * 
 * Handles enrollment operations including starting enrollment sessions.
 */

import axios from "axios";
import { useAuthStore } from "../store/authStore";

// import axios from '@/lib/api/axios-instance';
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8002';

export type EnrollmentStatus = 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled';

export interface EnrollmentStartRequest {
  student_id: number;
  device_id: number;
  finger_id: number;
}

export interface EnrollmentStartResponse {
  session_id: string;
  student_id: number;
  device_id: number;
  finger_id: number;
  status: EnrollmentStatus;
  started_at: string;
}

export interface EnrollmentCountResponse {
  successful_enrollments: number;
}

export interface ApiError {
  detail?: string;
  message?: string;
  code?: string;
  field_errors?: Record<string, string[]>;
}

/**
 * Custom error class for enrollment API errors.
 */
export class EnrollmentApiError extends Error {
  constructor(
    message: string,
    public statusCode: number,
    public code?: string,
    public fieldErrors?: Record<string, string>
  ) {
    super(message);
    this.name = 'EnrollmentApiError';
  }
}

/**
 * Start fingerprint enrollment for a student on a device.
 * 
 * This creates an enrollment session and sends the enrollment command to the device.
 * The device will enter enrollment mode and wait for the student to place their finger.
 * 
 * @param request - Enrollment start request with student_id, device_id, and finger_id
 * @returns Promise resolving to enrollment start response with session information
 * @throws EnrollmentApiError if request fails
 */
export async function startEnrollment(
  request: EnrollmentStartRequest
): Promise<EnrollmentStartResponse> {
  try {

    const authStore = useAuthStore.getState();
    const token = authStore.token;

    const response = await axios.post<EnrollmentStartResponse>(
      `${API_BASE_URL}/api/v1/enrollment/start`,
      request,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      }
    );

    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const statusCode = error.response?.status || 500;
      const errorData = error.response?.data as ApiError | undefined;

      // Handle field errors
      if (errorData?.field_errors) {
        const fieldErrors: Record<string, string> = {};
        Object.entries(errorData.field_errors).forEach(([field, messages]) => {
          fieldErrors[field] = messages.join(', ');
        });

        const message = errorData?.detail || errorData?.message || 'Validation error';
        throw new EnrollmentApiError(message, statusCode, errorData?.code, fieldErrors);
      }

      // Handle structured error responses (detail may be { message, code })
      if (errorData?.detail || errorData?.message) {
        const raw = errorData.detail || errorData.message;
        const message =
          typeof raw === "object" && raw && "message" in raw
            ? String((raw as { message: string }).message)
            : String(raw || `Failed to start enrollment (${statusCode})`);
        const code =
          typeof raw === "object" && raw && "code" in raw
            ? (raw as { code?: string }).code
            : errorData.code;
        throw new EnrollmentApiError(message, statusCode, code);
      }

      // Handle HTTP status codes with friendly messages
      if (statusCode === 404) {
        throw new EnrollmentApiError('Student or device not found', statusCode, 'NOT_FOUND');
      }
      if (statusCode === 503) {
        throw new EnrollmentApiError('Device is offline or unreachable', statusCode, 'DEVICE_OFFLINE');
      }
      if (statusCode === 409) {
        throw new EnrollmentApiError('Enrollment already in progress', statusCode, 'ENROLLMENT_IN_PROGRESS');
      }
      if (statusCode === 400) {
        throw new EnrollmentApiError(errorData?.detail || 'Invalid request', statusCode, errorData?.code);
      }

      const message = error.message || `Failed to start enrollment (${statusCode})`;
      throw new EnrollmentApiError(message, statusCode);
    }

    if (error instanceof EnrollmentApiError) {
      throw error;
    }

    throw new EnrollmentApiError('An unexpected error occurred', 500);
  }
}

/**
 * Get list of finger IDs already enrolled for a student on a device.
 *
 * @param deviceId - Device ID
 * @param studentId - Student ID
 * @returns Enrolled finger IDs (0-9)
 */
export async function getEnrolledFingers(
  deviceId: number,
  studentId: number
): Promise<{ device_id: number; student_id: number; finger_ids: number[] }> {
  const authStore = useAuthStore.getState();
  const token = authStore.token;

  const response = await axios.get<{ device_id: number; student_id: number; finger_ids: number[] }>(
    `${API_BASE_URL}/api/v1/enrollment/devices/${deviceId}/students/${studentId}/fingers`,
    {
      headers: { Authorization: `Bearer ${token}` },
    }
  );
  return response.data;
}

/**
 * Delete a fingerprint template for a student on a device.
 *
 * @param deviceId - Device ID
 * @param studentId - Student ID
 * @param fingerId - Finger index (0-9)
 */
export async function deleteFingerprint(
  deviceId: number,
  studentId: number,
  fingerId: number
): Promise<void> {
  const authStore = useAuthStore.getState();
  const token = authStore.token;

  await axios.delete(
    `${API_BASE_URL}/api/v1/enrollment/devices/${deviceId}/students/${studentId}/fingers/${fingerId}`,
    {
      headers: { Authorization: `Bearer ${token}` },
    }
  );
}

/**
 * Cancel an ongoing enrollment session.
 *
 * Sends cancel command to the device. The backend broadcasts the cancellation
 * via WebSocket so the frontend receives the update.
 *
 * @param sessionId - Enrollment session ID from start response
 * @throws EnrollmentApiError if request fails
 */
export async function cancelEnrollment(sessionId: string): Promise<void> {
  try {
    const authStore = useAuthStore.getState();
    const token = authStore.token;

    await axios.post(
      `${API_BASE_URL}/api/v1/enrollment/cancel/${encodeURIComponent(sessionId)}`,
      {},
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      }
    );
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const statusCode = error.response?.status || 500;
      const errorData = error.response?.data as ApiError | undefined;

      if (errorData?.detail || errorData?.message) {
        const message = errorData.detail || errorData.message || `Failed to cancel enrollment (${statusCode})`;
        throw new EnrollmentApiError(message, statusCode, errorData?.code);
      }
      throw new EnrollmentApiError(error.message || `Failed to cancel enrollment (${statusCode})`, statusCode);
    }
    throw new EnrollmentApiError('An unexpected error occurred', 500);
  }
}

/**
 * Get successful enrollment count for dashboard statistics.
 *
 * @param token - JWT authentication token
 * @returns Promise resolving to successful enrollment count
 */
export async function getSuccessfulEnrollmentCount(token: string): Promise<number> {
  const response = await axios.get<EnrollmentCountResponse>(
    `${API_BASE_URL}/api/v1/enrollment/stats/count`,
    {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
    }
  );

  return response.data.successful_enrollments;
}
